### PR TITLE
stream/okx: allow rate limit definitions to be used by the stream package

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

image

This allows for rate limits to be defined across connections if needed. 

OKx: 
* Share REST trading endpoint limits so that the websocket system can use them and not 429 our lives away.

Request:
* Add function GetRateLimiterDefinitions() to export request defined limiters to be shared across websocket implementation if needed.

stream:
* Allow the setting of a global rate limiting set across all potential connections which is backwards compatible
* Add param epl request.EndpointLimit to outbound request functions so that it can be matched to a specific rate limiter 
* Add commentary to Connection interface funcs
* Adjust writToConn function and remove secondary IsConnected() check as it will just error at a writeconn level anyway when using a closed connection. Also add coverage.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/stream/websocket_types.go [hunk 1]
```diff
@@ -121,6 +125,13 @@ type WebsocketSetup struct {
 	// MaxWebsocketSubscriptionsPerConnection defines the maximum number of
 	// subscriptions per connection that is allowed by the exchange.
 	MaxWebsocketSubscriptionsPerConnection int
+
+	// RateLimitDefinitions contains the rate limiters shared between WebSocket and REST connections for all endpoints.
+	// These rate limits take precedence over any rate limits specified in individual connection configurations.
+	// If no connection-specific rate limit is provided and the endpoint does not match any of these definitions,
+	// an error will be returned. However, if a connection configuration includes its own rate limit,
+	// it will fall back to that configuration’s rate limit without raising an error.
+	RateLimitDefinitions request.RateLimitDefinitions
```
Comment (gloriousCode) [inline exchanges/stream/websocket_types.go:134]: I was just thinking, why not use the requester itself? Check it out here
https://gist.github.com/gloriousCode/15e1e8a54cdbafec7add64e6cad183e4
Not a big deal though

File: exchanges/stream/websocket_connection.go [hunk 2]
```diff
@@ -67,52 +73,60 @@ func (w *WebsocketConnection) SendJSONMessage(ctx context.Context, data interfac
...
 		if w.Verbose {
 			log.Debugf(log.WebsocketMgr, "%v websocket connection: sending message [%s]\n", w.ExchangeName, message)
 		}
 		return w.Connection.WriteMessage(messageType, message)
 	})
 }
 
-func (w *WebsocketConnection) writeToConn(ctx context.Context, writeConn func() error) error {
+func (w *WebsocketConnection) writeToConn(ctx context.Context, epl request.EndpointLimit, writeConn func() error) error {
 	if !w.IsConnected() {
```
Comment (gloriousCode) [inline exchanges/stream/websocket_connection.go:86]: With the other removal, we could put this after the locks. The other one I'd argue is more important

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Good stuff. Two minor bits

File: exchanges/stream/websocket_types.go (hunk 1 above)
Comment (shazbert) [reply exchanges/stream/websocket_types.go:134]: Yeah that isn't a bad idea, the current diff though, doesn't look like it will be backwards compatible did it pass tests? error: corrupt patch at line 171 🪦 for me 😭. 

I personally prefer it being segregated, instead of relying on a correctly configured Requester, the okx implementation seems like an edge case, most definitions will be done per connection from what I quickly looked at, so keeping it as a more generalised component might be better. If I ever use the stream package for whatever reason to make a custom websocket link with no REST implementation point of view.

File: exchanges/stream/websocket_connection.go (hunk 2 above)
Comment (shazbert) [reply exchanges/stream/websocket_connection.go:86]: Yeah I did think about this, the only reason I had for keeping the upper one is so routines don't hang consecutively on the rate limit sleep when there is no available connection. When you are writing to the connection it does the check anyway and determines if its connected 🤷 and it's hard to get coverage for it trying to time execution. I won't resolve it just yet, maybe there is something else to consider by other reviews.

File: exchanges/stream/websocket_types.go (hunk 1 above)
Comment (gloriousCode) [reply exchanges/stream/websocket_types.go:134]: Yeah that reasoning is fine for me. I was thinking along the lines of you already have a requester by that point and it slightly simplified a bit of your code, but you're right in that it increased a dependency which was unnecessary

**Review by gloriousCode — APPROVED** [review]
tACK!

**Review by thrasher- — APPROVED** [review]
Changes ACK. Was curious about the PingHandler requiring a rate limit as I thought it would be excluded from requests however it's probably best to leave support for it just in a specific exchange limits it
